### ISSUE TITLE
Investigate Waffle vs Truffle

### ISSUE THREAD

**Issue by boolafish** [omisego/plasma-contracts#103]
Personally I'd love to have truffle on waffle~

### Background
It seems like py-solc is not really updating to support latest solidity. We can update that but whether we want to maintain that ourselves is a question. It seems like a good timing to move away. 

Also, related to previous audit recommendation: https://www.pivotaltracker.com/story/show/161138286

So one obvious option is truffle. However, Pepesza points out there's another framework called Waffle might be an alternative.

Some links for Waffle:
https://github.com/EthWorks/Waffle
https://medium.com/ethworks/meet-waffle-8e630e5e9e91

### What to do here
Let's put some reasoning, pros and cons on the two choices. Ideally make decision.

**Comment by kasima** [comment]
@jbunce – relevant to your interests.

(Jake would buy you a 🍩 if you decoupled contract deployment from elixir-omg.)

**Comment by kevsul** [comment]
**Truffle** provides tooling for compiling, deploying and testing smart contracts. Its deployment toolset allows you to deploy to any Ethereum network (local, Rinkeby, mainnet, etc.). One nice feature about it is that tests can be written in Solidity as well as Javascript.

**Waffle's** focus is on testing. It can compile smart contracts but not deploy them. All tests are run against a bundled Ganache. It provides some Chai extensions to make tests easier to write e.g. `expect(contract.method).to.emit(contract.Event)` 

**eth-tester** is the Python testing framework that we currently use. It does not provide tooling for either compiling or deploying smart contracts. Currently we use py-solc for compiling, but development on it appears to have halted and it's not compatible with Solidity 0.5

Any of the three tools can handle testing. However if we want to continue writing tests in Python, then eth-tester will probably need to be combined with one of the other two to compile Solidity 0.5 contracts.

I think it's a separate issue - but the possibility of using Truffle instead of dev_helpers to deploy the RootChain contracts has been raised. If that goes ahead, then it seems sensible to use Truffle here as well (instead of Waffle), at least for compiling the smart contracts.

I don't have a string opinion about writing tests in Javascript or Python. On one hand javascript is better integrated with Truffle and would be easier to setup and maintain. On the other hand there are already a lot of tests written in Python as well as experienced Python (and javascript averse) devs on the team…

TLDR: I think it makes more sense to go with Truffle rather than Waffle for compilation and deployment.
What the tests are written in depends on whether the Pythonistas want to learn Javascript or not 😬

**Comment by kevsul** [comment]
> I think it's a separate issue - but the possibility of using Truffle instead of dev_helpers to deploy the RootChain contracts has been raised

@jbunce @kasima I don't think a separate issue has been created for this yet, but I do think it belongs in elixir-omg rather than here, as it deals with replacing the current `dev_helpers` stuff with truffle. I'm open to being convinced otherwise though...

**Comment by pdobacz** [comment]
Some comments:

 - a note on Python and testing: AFAIK, Python + testing on `ganache` (getting rid of `eth-tester` itself) is an option too, and has been considered and postponed. It is however then reliant on `web3.py` whose status/future I'm not up-to-date about.
I'd favor getting rid of `eth-tester` in tests but remaining in Python-world, in general. But if the js-tooling is orders of magnitude better than Python's then, sigh, ok :).
Notwithstanding, there's a lot of tests there...
 - "the possibility of using Truffle instead of dev_helpers" - a reminder - truffle or not, `dev_helpers` **shouldn't be used** for anything production-like in the long run. State we're in now has always been an interim workaround to get ramped up with deployments. It is intended to be used by `elixir-omg`-specific integration tests.
Having said that, there still is an option of leveraging Truffle in `dev_helpers` somehow, but let's treat it separately from what Jake needs to roll his deployments conveniently. I think the technical requirements for those two tools diverge substantially.

**Comment by jbunce** [comment]
@kevsul I've opened https://github.com/omisego/elixir-omg/issues/734 for discussion

**Comment by boolafish** [comment]
Note from abstract layer migration plan meeting:
1. Waffle is basically out.
1. Let's separate the concern/issue of deployment and testing
1. Main concern on moving tests to truffle is on team's knowledge. Before the team is familiar with this new testing framework/language, the migration of the test suite might mean people not familiar on the assumption of them. And this might decrease the security in this sense.
1. Let's have a small call compare js testing POC and python to decide
1. in-house solution vs more standard solution

**Comment by boolafish** [comment]
My personal 2 yen on testing in general:

Python + Web3.py + Ganache:
- Python is a small plus (personally prefer python, but js with async/await + class, fairly speaking, is not that much different in coding experience)
- Web3.py is a minus (it is in `jsCodeStyle` instead of `python_code_style` + from experience usually not as updated as web3.js)
- Ganache setup is a small minus, slightly more complex to setup. (Need to run a process in beneath on the starting of test and pipe out the std output to remove noise messages, and, kill the process afterward)

JS + Truffle:
- Easier setup, fairly straight forward to write and run
- I think Mocha is slightly better than pytest in test description (customized sentence vs function name)

If people are curious how js test code looks like: [JS example test code](https://github.com/pgebal/plasma-contracts/blob/a660a81fe2be8a27de7522006ce88f54bc6992f6/abstract_poc/test/testPlasmaFramework.js#L52) by @pgebal 

re: Security-wise concern:
1. Short term, yes, if our team is not familiar with these new tools. However, long term speaking, consider the (sad) fact there is just more tooling in js rather than python in blockchain world, when we get any new hirings, the benefit of this just keep decreases. Also, this means we get less support community-wise.
1. I am pretty confident we can provide as good quality tests in js than in python. It is more about a matter of time and effort.

---
Again, I like python more than js. However, I am bringing this out because I think testing with truffle is a more widely used practice and the js toolings would bring us long term value (and sadly web3.py is just....a really badly written lib IMO). Also, I see this as probably our last chance to make this switch. With the huge change of abstract layer, the extra effort to switch might be justify-able at this moment. As more and more tests are added, the switch would just become a less feasible choice in the future.

**Comment by kevsul** [comment]
I totally agree with @boolafish 
I also prefer Python, but modern javascript has copied so much from Python that it's almost pleasant to use nowadays...

Truffle is very actively maintained and keeps improving. I know that there's an initial cost to learning it, but using it saves us from spending resources on our homegrown test framework.

**Comment by kasima** [comment]
What was the reason for throwing out waffle? Deployment tooling?

**Comment by boolafish** [comment]
@kasima yeah, it does not help deploy, and we probably don't want to have two new tools (like deploy in truffle and test in waffle).

**Comment by InoMurko** [comment]
Is the jury still out? The latest is a fight (what I gather from @boolafish ) 
Python + Web3.py + Ganache  *or*  JS + Truffle?

I think that with a bit of help we could do the deployment part because we need it most. But we definitely need a decision asap. 
cc @jbunce

**Comment by Pongch** [comment]
Hi I'd like to give my 2 cents as well. I've had a lot of discussions with the ODP people. A lot of dApp + tooling level developers seems to have preferences for `truffle.js` and have even asked a few times for a way to deploy the plasma contract in either a container or a deployment tool that is decoupled from the codebase. In both of these instances- their stack is JavaScript. 

I would argue that 90% of smart contracts deployment process is done with Truffle.

**Comment by boolafish** [comment]
Note for Action item in the meeting:
- [ ] Collect how much time needed to migrate the tests to js.

**Comment by boolafish** [comment]
[Notes of the meeting]
- Port existing python with [py-solcx](https://github.com/iamdefinitelyahuman/py-solc-x) that supports existing tests
- Write tests in each PR with JS. Implementer and code reviewer should double check with existing python test case. We would try to port existing code to js in this flavor on the way of building it.
